Feature: event backbone

jniles opened the pull request:
This PR implements Topic's publish methods throughout the as many stable server controllers as possible.  It also improves error handling on the client in case of network loss, increasing the retry poll time to 10 seconds.

---

Thank you for contributing!

Before submitting this pull request, please verify that you have:
- [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

jniles committed:
chore(events): implement events through app

This commit implements Topic's publish methods throughout the as many
stable server controllers as possible.  These events will later be
handled on the client via an EventStream.

jniles committed:
fix(events): client side event error handling

This commit implements a longer retry interval for polling for an
internet connection.  It also implements error handlers on the
EventStream created on the client.

sfount commented on client/src/js/services/system.js:
```diff
@@ -34,9 +34,22 @@ function SystemService($http, util) {
   }
 
+  function handleErrorEvent(event) {
+    if (event.readyState == EventSource.CLOSED) {
```
Is there any reason this isn't strict equals?

jniles replied:
Whoops.  Nope - I'll address that.

sfount commented on server/controllers/admin/employees.js:
```diff
@@ -44,30 +53,31 @@ exports.list = function (req, res, next) {
 exports.listHolidays = function (req, res, next) {
-  var pp = JSON.parse(req.params.pp);
-  var sql =
+  let pp = JSON.parse(req.params.pp);
```
Do you know what `pp` means? Not a big issue but this should be discouraged in the future or very well commented.

jniles replied:
Noted.  I have no idea what `pp` is.  I think it's something like 'parameters'.

sfount commented on server/controllers/finance/cashboxes/currencies.js:
```diff
@@ -0,0 +1,150 @@
+
+  db.exec(sql, [req.params.id, req.params.currencyId])
+  .then(function (rows) {
+    if (!rows.length) {
```
We have a lot of different ways of doing this on the server, some examples; 

``` js
if (rows.length === 0) { //...
if (!rows.length) { //...

// Contrived 
let LENGTH_EMPTY = 0;
if (rows.length === LENGTH_EMPTY) { //...

let isEmpty = rows.length === 0;
if (isEmpty) { //...
```

I suggest that we tackle trying to make this uniform now before server business logic becomes more complex going forward.

Something like

``` js
if (utility.isEmpty(databaseQueryResult)) { //...
```

What do you think?
@IMA-WorldHealth/local-contributors

jniles replied:
We could always use `lodash`.

``` js
if (_.isEmpty(rows)) {
  // do something
}
```

I'd be comfortable with lodash myself.

sfount commented on server/controllers/admin/suppliers.js:
```diff
@@ -66,26 +78,50 @@ function detail(req, res, next) {
 
   // provide uuid if the client has not specified
-  data.uuid = db.bid(data.uuid || uuid());
+  const uid = data.uuid || uuid();
```
here the `uuid` library is directly invoked with `uuid()` - in another file (`location.js` I think) `uuid.v4()` is used to retrieve a uuid. Should we stick to one of these methods?

sfount replied:
Briefly reading the `node-uuid` source code shows that `uuid.v4` is assigned as the library export - this means both provide the same output but it may still be a good idea to stick to one or another for the sake of future developers.

EDIT

``` js
var uuid = v4;
uuid.v1 = v1;
uuid.v4 = v4;

// ...
module.exports = uuid;
```

jniles replied:
The `uuid()` method of calling a version 4 uuid [is deprecated](https://github.com/broofa/node-uuid#deprecated-apis).  We should stick to `uuid.v4()`.  I'm not sure how this got through the initial review, but I"ll update it!

jniles committed:
fix(review): implement review corrections

This commit implements feedback from @sfount on the PR.  The server was
also scanned for errant `uuid()` calls instead of `uuid.v4()`.
